Sources/NIOCore/NIOIPAddresses.swift

```diff
@@ -0,0 +1,583 @@
+        switch bytes.count {
+        case 4: self = .v4(try .init(packedBytes: bytes))
+        case 16: self = .v6(try .init(packedBytes: bytes))
+        default:
+            throw IPAddressError.bytesArrayHasWrongLength(bytes.count)
+        }
+    }
+    
+    /// Creates a new `NIOIPAddress` for the given libc IPv4 address.
+    ///
+    /// - parameters:
+    ///     - posixIPv4Address: libc ipv4 address.
+    public init(posixIPv4Address: in_addr) {
+        let uint8AddressBytes: IPv4Bytes = .init((
+            UInt8(truncatingIfNeeded: posixIPv4Address.s_addr >> 24),
+            UInt8(truncatingIfNeeded: posixIPv4Address.s_addr >> 16),
+            UInt8(truncatingIfNeeded: posixIPv4Address.s_addr >> 8),
+            UInt8(truncatingIfNeeded: posixIPv4Address.s_addr)
+        ))
+        
+        self = .v4(.init(address: uint8AddressBytes))
+    }
+    
+    /// Creates a new `NIOIPAddress` for the given libc IPv6 address.
+    ///
+    /// - parameters:
+    ///     - posixIPv6Address: libc ipv6 address.
+    public init(posixIPv6Address: in6_addr) {
+        self = .v6(.init(address: .init(posixIPv6Address.__u6_addr.__u6_addr8)))
+    }
+}
+
+extension UInt8 {
+    /// Creates UInt8 from hex character in form of an ascii value.
+    init?(hexAsciiValue: UInt8) {
+        switch hexAsciiValue {
+        case UInt8(ascii: "0"): self = 0
+        case UInt8(ascii: "1"): self = 1
+        case UInt8(ascii: "2"): self = 2
+        case UInt8(ascii: "3"): self = 3
+        case UInt8(ascii: "4"): self = 4
+        case UInt8(ascii: "5"): self = 5
+        case UInt8(ascii: "6"): self = 6
+        case UInt8(ascii: "7"): self = 7
+        case UInt8(ascii: "8"): self = 8
+        case UInt8(ascii: "9"): self = 9
+        case UInt8(ascii: "a"): self = 10
+        case UInt8(ascii: "A"): self = 10
+        case UInt8(ascii: "b"): self = 11
+        case UInt8(ascii: "B"): self = 11
+        case UInt8(ascii: "c"): self = 12
+        case UInt8(ascii: "C"): self = 12
+        case UInt8(ascii: "d"): self = 13
+        case UInt8(ascii: "D"): self = 13
+        case UInt8(ascii: "e"): self = 14
+        case UInt8(ascii: "E"): self = 14
+        case UInt8(ascii: "f"): self = 15
+        case UInt8(ascii: "F"): self = 15
+        default: return nil
+        }
+    }
+}
+
+extension UInt32 {
+    /// Creates an integer from the given UInt8 tuple.
+    init(uint8Tuple: (UInt8, UInt8, UInt8, UInt8)) {
+        self = UInt32(uint8Tuple.0) << 24
+            + UInt32(uint8Tuple.1) << 16
+            + UInt32(uint8Tuple.2) << 8
+            + UInt32(uint8Tuple.3)
+    }
+}
+
+extension UInt16 {
+    /// Creates an integer from the given `UInt8` tuple.
+    init(uint8Tuple: (UInt8, UInt8)) {
+        self = UInt16(uint8Tuple.0) << 8
+             + UInt16(uint8Tuple.1)
+    }
+    
+    /// Interprets value as hex and creates ascii `UInt8`.
+    var asciiValue: UInt8 {
```

Comment:
There's absolutely no reason to use unsafe code for this.

Reply:
The `ASCII(_unsafeUnchecked: ...)` is an internal WebURL API which is not "unsafe" as I think you're thinking of "unsafe".

It's called "unsafe" to draw attention to the fact that it does not validate that values are <128. There are reasons (beyond the scope of this PR) why it could be problematic for such values to exist, and why additional precautions are necessary when using that API, to ensure they are never created (e.g. creating a value with a byte taken from a string literal, as here, is safe and doesn't need a run-time check).

Reply:
Granted, that makes sense. I agree this can be faster, but again I'd like to land this PR first and iterate the performance later.